CRM/Contribute/BAO/Contribution.php

```diff
@@ -2438,9 +2438,14 @@ public function composeMessageArray(&$input, &$ids, &$values, $recur = FALSE, $r
     if (empty($this->_component)) {
       $this->_component = CRM_Utils_Array::value('component', $input);
     }
+    if (isset($input['is_email_receipt'])) {
+      $values['is_email_receipt'] = $input['is_email_receipt'];
+    }
 
+    // Store values in a temp variable, so they don't get lost after _gatherMessageValues
+    $tmp_values = $values;
     //not really sure what params might be passed in but lets merge em into values
-    $values = array_merge($this->_gatherMessageValues($input, $values, $ids), $values);
+    $values = array_merge($this->_gatherMessageValues($input, $values, $ids), $tmp_values);
     if (!empty($input['receipt_date'])) {
       $values['receipt_date'] = $input['receipt_date'];
     }
```

Comment:
Let's ditch this bit because I feel it's blocking the PR. That function is only called from here & ideally it would be dismantled into the calling function in bits e.g

extract this

```
    if ($this->address_id) {
      $addressParams = array('id' => $this->address_id);
      $addressDetails = CRM_Core_BAO_Address::getValues($addressParams, FALSE, 'id');
      $addressDetails = array_values($addressDetails);
    }
    // Else we assign the billing address of the contribution contact.
    else {
      $addressParams = array('contact_id' => $this->contact_id, 'is_billing' => 1);
      $addressDetails = (array) CRM_Core_BAO_Address::getValues($addressParams);
      $addressDetails = array_values($addressDetails);
    }

    if (!empty($addressDetails[0]['display'])) {
      $values['address'] = $addressDetails[0]['display'];
    }
```
into a function & call from composeMessageArray, ideally with tests, and possibly some review of how consistently that address token is used.

slow & painful, yes.